test/Writer/Factory/WriterFactoryTest.php

```diff
@@ -0,0 +1,290 @@
+        $object = $factory->createService($pluginManager, InvokableObject::class, InvokableObject::class);
+
+        // Assert
+        $this->assertInstanceOf(InvokableObject::class, $object);
+        $this->assertEquals([], $object->options);
+    }
+
+    /**
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::__construct
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::createService
+     */
+    public function testCreateServiceWithCreationOptions()
+    {
+        // Arrange
+        $container = $this->getMockForAbstractClass(ServiceLocatorInterface::class);
+
+        $pluginManager = $this->prophesize(AbstractPluginManager::class);
+        $pluginManager->getServiceLocator()->willReturn($container);
+        $pluginManager = $pluginManager->reveal();
+
+        $factory = new WriterFactory(['foo' => 'bar']);
+
+        // Act
+        $object = $factory->createService($pluginManager, InvokableObject::class, InvokableObject::class);
+
+        // Assert
+        $this->assertInstanceOf(InvokableObject::class, $object);
+        $this->assertEquals(['foo' => 'bar'], $object->options);
+    }
+
+    /**
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::__construct
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::createService
+     */
+    public function testCreateServiceWithValidRequestName()
+    {
+        // Arrange
+        $container = $this->getMockForAbstractClass(ServiceLocatorInterface::class);
+
+        $pluginManager = $this->prophesize(AbstractPluginManager::class);
+        $pluginManager->getServiceLocator()->willReturn($container);
+        $pluginManager = $pluginManager->reveal();
+
+        $factory = new WriterFactory();
+
+        // Act
+        $object = $factory->createService($pluginManager, 'invalid', InvokableObject::class);
+
+        // Assert
+        $this->assertInstanceOf(InvokableObject::class, $object);
+        $this->assertEquals([], $object->options);
+    }
+
+    /**
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::__construct
+     * @covers \Zend\Log\Writer\Factory\WriterFactory::createService
+     * @expectedException \Zend\ServiceManager\Exception\InvalidServiceException
+     * @expectedExceptionMessage Zend\Log\Writer\Factory\WriterFactory requires that the requested
+     * name is provided on invocation; please update your tests or consuming container
+     */
+    public function testCreateServiceInvalidNames()
+    {
+        // Arrange
+        $container = $this->getMockForAbstractClass(ServiceLocatorInterface::class);
+
+        $pluginManager = $this->prophesize(AbstractPluginManager::class);
+        $pluginManager->getServiceLocator()->willReturn($container);
+        $pluginManager = $pluginManager->reveal();
+
+        $factory = new WriterFactory();
+
+        // Act
+        $object = $factory->createService($pluginManager, 'invalid', 'invalid');
+
+        // Assert
+        $this->assertInstanceOf(InvokableObject::class, $object);
```

Comment:
The exceptions will be raised before this.

This is one place where I wouldn't use the annotations, but instead call `setExpectedException()` immediately prior to the call that will raise the exception; this makes it clear when setup code raises an exception vs when execution code raises it. As such:

```php
$this->setExpectedException(
    InvalidServiceException::class,
    'WriterFactory requires that the requested name is provided'
);
$factory->createService($pluginManager, 'invalid', 'invalid');
```